lib/bom.gradle

```diff
@@ -21,7 +21,32 @@ configure(projectsWithFlags('bom')) {
                 }
                 return "${a.ext.artifactId}".compareTo("${b.ext.artifactId}")
             }.each { p ->
-                api "${p.group}:${p.ext.artifactId}:${p.version}"
+                if (rootProject.ext.has('bomGroups')) {
+                    def bomGroups = rootProject.ext.bomGroups
+                    if (!(bomGroups instanceof Map)) {
+                        throw new IllegalStateException("bomGroups must be a Map: ${bomGroups}")
+                    }
+
+                    if (!bomGroups.containsKey(':' + project.name)) {
+                        api "${p.group}:${p.ext.artifactId}:${p.version}"
```

Comment:
I added this logic in order to add dependencies to bom file when the user add `bomGroups` but don't add some project. It will make the bom file of the excluded project from bomGroups property have dependencies like that the user didn't use the `bomGroups` property.

Reply:
>I added this logic in order to add dependencies to bom file when the user add bomGroups property but exclude some project

I'm sorry that I don't understand what you meant here. 😅 
Could you elaborate more on this with an example, please? 🙏

Reply:
For example, let's use same case with PR's description except for `bomGroups` property. And let's say we add only module2 to the `bomGroups` property.
```
ext {
    bomGroups = [
            ':module2': [':module2:submodule1', ':module2:submodule2']
    ]
}
```

1. Without this logic
 Then even thought we set `bom` flag to `module1` in `setting.gradle`, module1's bom file has no dependencies. 
```
...
<!-- module1's bom file -->
<!-- No dependencies -->
...
```

2. With this logic
 we can see the all dependencies in `module1's bom file`.
```
...
<!-- module1's bom file -->
<dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>com.example.hy</groupId>
        <artifactId>gradle-scripts-test-module1-submodule1</artifactId>
        <version>0.0.1-SNAPSHOT</version>
      </dependency>
      <dependency>
        <groupId>com.example.hy</groupId>
        <artifactId>gradle-scripts-test-module1-submodule2</artifactId>
        <version>0.0.1-SNAPSHOT</version>
      </dependency>
      <dependency>
        <groupId>com.example.hy</groupId>
        <artifactId>gradle-scripts-test-module2-submodule1</artifactId>
        <version>0.0.1-SNAPSHOT</version>
      </dependency>
      <dependency>
        <groupId>com.example.hy</groupId>
        <artifactId>gradle-scripts-test-module2-submodule2</artifactId>
        <version>0.0.1-SNAPSHOT</version>
      </dependency>
    </dependencies>
  </dependencyManagement>
...
```

Reply:
Thanks for the explanation. 😄 
I get it. 😄 

So basically if a bom module is specified in `bomGroups`, it only contains the modules in that `bomGroups`.
If it's not it contains all module that have `publish` flags.

This all looks good. 👍 
However, I wonder if we can put all bom-related information into one place instead of having them in `settings.gradle` and `build.gradle`.

How about introducing a key concept and using it in `settings.gradle`?
```
includeWithFlags ':module1', 'bom:key1'
includeWithFlags ':module1-dependent', 'java', 'publish', 'key1'

includeWithFlags ':module2', 'bom:key2'
includeWithFlags ':module2-dependent', 'java', 'publish', 'key2'

includeWithFlags ':common', 'java', 'publish', 'key1', 'key2'
```

/cc @line/dx

Reply:
> How about introducing a key concept and using it in settings.gradle?

Actually, I thought that at the first. But I was worried that it would be difficult to distinguish between `key` and `flag`.

Reply:
> distinguish between key and flag. Any ideas to prevent a key with the same name as a flag?

I thought we could validate it in the `includeWithFlags` method.
We can check if the built-in flags are used as a key but not for a custom flag.

So probably, my idea was not a good idea. 😓 
Let's get back to the current approach.
The bom in `bomGroups` only includes the modules that are specified in the map.
The bom not in `bomGroups` includes all the modules that have `java` and `publish` flag. Is that right?

Reply:
>The bom in bomGroups only includes the modules that are specified in the map.
The bom not in bomGroups includes all the modules that have java and publish flag. Is that right?

Right!

Reply:
Yeah, let's do it as is. 😉